Getting started section (#323)
holoviz/lumen: docs/getting_started/build_an_app.md

# Build an App
This guide will guide you through the steps of how to build a Lumen app that will look like this:

![](../_static/getting_started/build_app_07.png)

## Building the app
This section will walk through the steps to creating the app.
Create a file named `penguins.yaml`.

> The first thing which is needed is a source. Here we will add a penguin source from a file that contains a penguin's table.

Reviewer: ```suggestion
The first thing which is needed is a source. Here we will add file source that will fetch data about individual penguins of three different and various measurement about them (such as the length of their bill).
```

Author: I will commit your suggestion, just want to note that the way I built up the sentences was to try to match the words in the sentence with what is happening in the YAML specification.

Then launch the app with `lumen serve penguins.yaml --autoreload` in your terminal and open `http://localhost:5006` in your browser of choice.
By using `--autoreload`, the dashboard automatically refreshes and updates the application whenever we make changes to the YAML file.

::::{tab-set}
:::{tab-item} YAML
:sync: yaml
``` {code-block} yaml

sources:
  penguin_source:
    type: file
    tables:
      table_penguin: https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-28/penguins.csv
```
:::

:::{tab-item} Preview
:sync: preview

![](../_static/getting_started/build_app_00.png)
:::
::::

This returns an empty dashboard because we don't have a targeted view.

> The simplest view to add is a table with the penguin's table in it.

Reviewer: Can you keep paragraph together rather than starting each sentence on a new line.

Author: I like having sentences separated by a new line; the output will be the same when converted to HTML. 

I think it has some advantages (even though they are small):
- For version control, it is easier to see what sentences have changed.  
- When staging changes on a large paragraph with multiple changes across sentences, I have a tendency to either accept or discard all the changes, new lines give me more fine-grained control. 
- When using spell checking, I can easily see what sentence has a spelling error based on the line.

Reviewer 2: Those are all good points @Hoxbro though I see a few downsides too:

1. This style hurts the fluidity of readability when looking at the Markdown file.
2. Is this something we would try to enforce throughout or is this something you only want to do for yourself when working on the docs?

As I think this would be hard to enforce throughout, it would be fine with me if @Hoxbro uses this approach for himself as long as there is no expectation that other people will follow this style (and people could rewrap the text if they want). The downside with *that* is inconsistency...

My only other suggestion might be that this style of separating sentences might be a good idea while working on a docs PR but we should rewrap the text in a consistent way before merge. How does that sound?

Reviewer: Right readability and consistency are my main concerns too. I agree with all your points @Hoxbro but if we decide to do this then I think we should do it consistently and it just hurts my eyes a little too much to agree to that.

Reviewer: Let's get @droumis' take here as well.

Reviewer 2: So for consistency, should we agree on a rewrap width before merging docs PRs? Before the merge, I'm ok with any style...

Author: I don't think it needs to be as black and white as both of you make it out to be. I think having different "formats" across the markdown files is fine, as long as the final product (the HTML) is consistent.  

For readability, I don't think it is as big of an eyesore, it is different but you get used to it quickly.

Reviewer: Leaving it for now but still leaning towards wrapping the paragraphs.

::::{tab-set}
:::{tab-item} YAML
:sync: yaml
``` {code-block} yaml

> :emphasize-lines: 7-12

Reviewer: The emphasis use is great!

targets:
  - title: Penguins
    source: penguin_source
    views:
      - type: table
        table: penguin_table

```
:::

:::{tab-item} Preview
:sync: preview

![](../_static/getting_started/build_app_01.png)
:::
::::

The table gives a good understanding of the data, but to really understand it, we have to visualize it.
This can be done with a [hvplot](https://hvplot.holoviz.org/) view and is as easy as replacing the table type with a hvplot type.

::::{tab-set}
:::{tab-item} YAML
:sync: yaml
``` {code-block} yaml
:emphasize-lines: 11

sources:
  penguin_source:
    type: file
    tables:
      table_penguin: https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-28/penguins.csv

targets:
  - title: Penguins
    source: penguin_source
    views:
      - type: hvplot
        table: penguin_table
```
:::

:::{tab-item} Preview
:sync: preview

![](../_static/getting_started/build_app_02.png)
:::
::::

This plot is a bit overwhelming, so instead of plotting everything in one plot, we can plot `bill_length_mm` on the x-axis and `bill_depth_mm` on the y-axis.

> Furthermore, we can color based on the specie and change the kind of the plot to scatter.

Reviewer: ```suggestion
Furthermore, we can color based on the species and change the kind of the plot to scatter.
```

        kind: scatter
        color: species
```
:::

:::{tab-item} Preview
:sync: preview

![](../_static/getting_started/build_app_04.png)
:::
::::

We can even expand the views with more plots like histograms of some of the penguin's features.

::::{tab-set}
:::{tab-item} YAML
:sync: yaml

``` {code-block} yaml
:emphasize-lines: 26-34

sources:
  penguin_source:
    type: file
    tables:
      penguin_table: https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-28/penguins.csv

pipelines:
  penguin_pipeline:
    source: penguin_source
    table: penguin_table
    filters:
      - type: widget
        field: sex
      - type: widget
        field: island

targets:
  - title: Penguins
    pipeline: penguin_pipeline
    views:
      - type: hvplot
        x: bill_length_mm
        y: bill_depth_mm
        kind: scatter
        color: species
      - type: hvplot
        kind: hist
        y: bill_length_mm
      - type: hvplot
        kind: hist
        y: bill_depth_mm
      - type: hvplot
        kind: hist
        y: body_mass_g

```
:::

:::{tab-item} Preview
:sync: preview

![](../_static/getting_started/build_app_05.png)
:::
::::

> The layout of the views is in a column where we can't even see the last plot, and the dashboard is non-responsive.

Reviewer: ```suggestion
The default layout we get is less than ideal for this case since it cuts off one of our plots, leaves a lot of empty space and does not resize responsively.
```
